ibm_cloud_sdk_core/base_service.py

```
@@ -23,134 +23,72 @@
-                if self.PASSWORD in self.vcap_service_credentials:
-                    self.password = self.vcap_service_credentials.get(self.PASSWORD)
-                if self.APIKEY in self.vcap_service_credentials:
-                    self.set_iam_apikey(self.vcap_service_credentials.get(self.APIKEY))
-                if self.IAM_APIKEY in self.vcap_service_credentials:
-                    self.set_iam_apikey(self.vcap_service_credentials.get(self.IAM_APIKEY))
-                if self.IAM_ACCESS_TOKEN in self.vcap_service_credentials:
-                    self.set_iam_access_token(self.vcap_service_credentials.get(self.IAM_ACCESS_TOKEN))
-                if self.ICP4D_URL in self.vcap_service_credentials:
-                    self.icp4d_url = self.vcap_service_credentials.get(self.ICP4D_URL)
-                if self.ICP4D_ACCESS_TOKEN in self.vcap_service_credentials:
-                    self.set_icp4d_access_token(self.vcap_service_credentials.get(self.ICP4D_ACCESS_TOKEN))
-
-        if (self.username is None or self.password is None) and self.token_manager is None:
-            raise ValueError(
-                'You must specify your IAM api key or username and password service '
-                'credentials (Note: these are different from your IBM Cloud id)')
+            if vcap_service_credentials is not None and isinstance(
+                    vcap_service_credentials, dict):
+                if vcap_service_credentials.get('username') and vcap_service_credentials.get('password'): # cf
+                    vcap_service_credentials['auth_type'] = 'basic'
+                elif vcap_service_credentials.get('apikey'): # rc
+                    vcap_service_credentials['auth_type'] = 'iam'
+                self._set_authenticator_properties(vcap_service_credentials)
+                self._set_service_properties(vcap_service_credentials)
+
+        if not self.authenticator:
+            self.authenticator = NoAuthAuthenticator()
```

We decided that an `authenticator` would be `required` in the base service. If one is not set, rather than use a default, we throw an error. In the `generated` service for Watson, we will specify the default of using the environment to create an authenticator. I think you need a separate module for this, since you user should also be able to get an authenticator like this without relying on the default.